ccaas_builder/cmd/build/main.go

```diff
@@ -0,0 +1,175 @@
+// connection.json file that is supplied in the
+// chaincode package.
+type Connection struct {
+	Address     string `json:"address"`
+	DialTimeout string `json:"dial_timeout"`
+	TLS         bool   `json:"tls_required"`
+	ClientAuth  bool   `json:"client_auth_required"`
+	RootCert    string `json:"root_cert"`
+	ClientKey   string `json:"client_key"`
+	ClientCert  string `json:"client_cert"`
+}
+
+type Config struct {
+	PeerName string
+}
+
+func main() {
+	logger.Println("::Build")
+
+	if err := run(); err != nil {
+		logger.Printf("::Error: %v\n", err)
+		os.Exit(1)
+	} else {
+		logger.Printf("::Type detected as ccaas")
+	}
+}
+
+func run() error {
+	sourceDir, metadataDir, outputDir := os.Args[1], os.Args[2], os.Args[3]
+	connectionSrcFile := filepath.Join(sourceDir, "/connection.json")
+	metadataFile := filepath.Clean(filepath.Join(metadataDir, "metadata.json"))
+	connectionDestFile := filepath.Join(outputDir, "/connection.json")
+	metainfoSrcDir := filepath.Join(sourceDir, "META-INF")
+	metainfoDestDir := filepath.Join(outputDir, "META-INF")
+
+	// Process and check the metadata file, then copy to the output location
+	metadataFileContents, err := ioutil.ReadFile(metadataFile)
+	if err != nil {
+		return err
+	}
+
+	var metadata ChaincodeMetadata
+	if err := json.Unmarshal(metadataFileContents, &metadata); err != nil {
+		return err
+	}
+
+	if strings.ToLower(metadata.Type) != "ccaas" {
+		return fmt.Errorf("chaincode type should be ccaas, it is %s", metadata.Type)
+	}
+
+	if err := copy.Copy(metadataDir, outputDir); err != nil {
+		return fmt.Errorf("failed to copy build metadata folder: %s", err)
+	}
+
+	if _, err := os.Stat(metainfoSrcDir); !os.IsNotExist(err) {
+		if err := copy.Copy(metainfoSrcDir, metainfoDestDir); err != nil {
+			return fmt.Errorf("failed to copy build META-INF folder: %s", err)
+		}
+	}
+
+	// Process and update the connections file
+	fileInfo, err := os.Stat(connectionSrcFile)
+	if err != nil {
+		return fmt.Errorf("connection.json not found in source folder: %s", err)
+	}
+
+	connectionFileContents, err := ioutil.ReadFile(connectionSrcFile)
+	if err != nil {
+		return err
+	}
+
+	// read the connection.json file into structure to process
+	var connectionData Connection
+	if err := json.Unmarshal(connectionFileContents, &connectionData); err != nil {
+		return err
+	}
+
+	// Treat each of the string fields in the connection.json as Go template
+	// strings. They can be fixed strings, but if they are templates
+	// then the JSON string that is defined in CHAINCODE_AS_A_SERVICE_BUILDER_CONFIG
+	// is used as the 'context' to parse the string
+
+	updatedConnection := Connection{}
+	var cfg map[string]interface{}
+
+	cfgString := os.Getenv("CHAINCODE_AS_A_SERVICE_BUILDER_CONFIG")
+	if cfgString != "" {
+		if err := json.Unmarshal([]byte(cfgString), &cfg); err != nil {
+			return fmt.Errorf("Failed to unmarshal %s", err)
+		}
+	}
+
+	updatedConnection.Address, err = execTempl(cfg, connectionData.Address)
```

Comment:
I'm not following what value the template use provides.

Reply:
Within each org the deployed chaincode package needs to be identical. Take the case of two peers in an org both with the same package (with then the same connection.json).  Unless this connection.json can be modified by the builder only a single instance of chaincode process can be used.  

Causes problems for scalability and availability.